File: dataset/CoT - Copy/CLAUDE_claude-sonnet-4-5-20250929/Task46.cpp

#include <iostream>
#include <string>
#include <regex>
#include <random>
#include <sstream>
#include <iomanip>
#include <sqlite3.h>
#include <openssl/sha.h>

class Task46 {
private:
    static const std::string DB_NAME;
    static const std::regex EMAIL_PATTERN;
    
    static std::string generateSalt() {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 255);
        
        std::stringstream ss;
        for (int i = 0; i < 16; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << dis(gen);
        }
        return ss.str();
    }
    
    static std::string hashPassword(const std::string& password, const std::string& salt) {
        std::string salted = salt + password;
        unsigned char hash[SHA256_DIGEST_LENGTH];
        SHA256(reinterpret_cast<const unsigned char*>(salted.c_str()), salted.length(), hash);
        
        std::stringstream ss;
        for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(hash[i]);
        }
        return ss.str();
    }
    
    static bool validateName(const std::string& name) {
        size_t len = name.length();
        return !name.empty() && len >= 2 && len <= 100;
    }
    
    static bool validateEmail(const std::string& email) {
        return !email.empty() && std::regex_match(email, EMAIL_PATTERN);
    }
    
    static bool validatePassword(const std::string& password) {
        size_t len = password.length();
        return !password.empty() && len >= 8 && len <= 128;
    }
    
    static std::string trim(const std::string& str) {
        size_t first = str.find_first_not_of(' ');
        if (first == std::string::npos) return "";
        size_t last = str.find_last_not_of(' ');
        return str.substr(first, (last - first + 1));
    }
    
    static std::string toLower(const std::string& str) {
        std::string result = str;
        for (char& c : result) {
            c = std::tolower(c);
        }
        return result;
    }

public:
    static void initializeDatabase() {
        sqlite3* db;
        char* errMsg = nullptr;
        
        int rc = sqlite3_open(DB_NAME.c_str(), &db);
        if (rc) {
            std::cerr << "Can't open database: " << sqlite3_errmsg(db) << std::endl;
            return;
        }
        
        const char* sql = "CREATE TABLE IF NOT EXISTS users ("
                         "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                         "name TEXT NOT NULL,"
                         "email TEXT UNIQUE NOT NULL,"
                         "password_hash TEXT NOT NULL,"
                         "salt TEXT NOT NULL)";
        
        rc = sqlite3_exec(db, sql, nullptr, nullptr, &errMsg);
        if (rc != SQLITE_OK) {
            std::cerr << "SQL error: " << errMsg << std::endl;
            sqlite3_free(errMsg);
        }
        
        sqlite3_close(db);
    }
    
    static std::pair<bool, std::string> registerUser(const std::string& name, 
                                                      const std::string& email, 
                                                      const std::string& password) {
        // Input validation
        if (!validateName(name)) {
            return {false, "Invalid name. Must be 2-100 characters."};
        }
        
        if (!validateEmail(email)) {
            return {false, "Invalid email format."};
        }
        
        if (!validatePassword(password)) {
            return {false, "Invalid password. Must be 8-128 characters."};
        }
        
        // Generate salt and hash password
        std::string salt = generateSalt();
        std::string passwordHash = hashPassword(password, salt);
        
        // Store in database
        sqlite3* db;
        int rc = sqlite3_open(DB_NAME.c_str(), &db);
        if (rc) {
            return {false, "Database connection failed."};
        }
        
        const char* sql = "INSERT INTO users (name, email, password_hash, salt) VALUES (?, ?, ?, ?)";
        sqlite3_stmt* stmt;
        
        rc = sqlite3_prepare_v2(db, sql, -1, &stmt, nullptr);
        if (rc != SQLITE_OK) {
            sqlite3_close(db);
            return {false, "Failed to prepare statement."};
        }
        
        std::string trimmedName = trim(name);
        std::string trimmedEmail = toLower(trim(email));
        
        sqlite3_bind_text(stmt, 1, trimmedName.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(stmt, 2, trimmedEmail.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(stmt, 3, passwordHash.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(stmt, 4, salt.c_str(), -1, SQLITE_TRANSIENT);
        
        rc = sqlite3_step(stmt);
        sqlite3_finalize(stmt);
        sqlite3_close(db);
        
        if (rc == SQLITE_DONE) {
            return {true, "User registered successfully."};
        } else if (rc == SQLITE_CONSTRAINT) {
            return {false, "Email already registered."};
        } else {
            return {false, "Registration failed."};
        }
    }
};

const std::string Task46::DB_NAME = "users.db";
const std::regex Task46::EMAIL_PATTERN("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\\\.[A-Za-z]{2,}$");

int main() {
    Task46::initializeDatabase();
    
    std::cout << "=== User Registration System Test Cases ===" << std::endl << std::endl;
    
    // Test Case 1: Valid registration
    auto result1 = Task46::registerUser("John Doe", "john.doe@example.com", "SecurePass123");
    std::cout << "Test 1 - Valid registration: " << result1.second << std::endl;
    
    // Test Case 2: Duplicate email
    auto result2 = Task46::registerUser("Jane Doe", "john.doe@example.com", "AnotherPass456");
    std::cout << "Test 2 - Duplicate email: " << result2.second << std::endl;
    
    // Test Case 3: Invalid email format
    auto result3 = Task46::registerUser("Bob Smith", "invalid-email", "Password789");
    std::cout << "Test 3 - Invalid email: " << result3.second << std::endl;
    
    // Test Case 4: Weak password
    auto result4 = Task46::registerUser("Alice Brown", "alice@example.com", "123");
    std::cout << "Test 4 - Weak password: " << result4.second << std::endl;
    
    // Test Case 5: Invalid name
    auto result5 = Task46::registerUser("A", "test@example.com", "ValidPass123");
    std::cout << "Test 5 - Invalid name: " << result5.second << std::endl;
    
    return 0;
}

File: dataset/CoT/CLAUDE_claude-sonnet-4-5-20250929/Task46.java

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.sql.*;
import java.util.Base64;
import java.util.regex.Pattern;

public class Task46 {
    private static final String DB_URL = "jdbc:sqlite:users.db";
    private static final Pattern EMAIL_PATTERN = Pattern.compile(
        "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\\\.[A-Za-z]{2,}$"
    );
    
    public static class RegistrationResult {
        public boolean success;
        public String message;
        
        public RegistrationResult(boolean success, String message) {
            this.success = success;
            this.message = message;
        }
    }
    
    private static void initializeDatabase() {
        try (Connection conn = DriverManager.getConnection(DB_URL);
             Statement stmt = conn.createStatement()) {
            String sql = "CREATE TABLE IF NOT EXISTS users (" +
                        "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                        "name TEXT NOT NULL," +
                        "email TEXT UNIQUE NOT NULL," +
                        "password_hash TEXT NOT NULL," +
                        "salt TEXT NOT NULL)";
            stmt.execute(sql);
        } catch (SQLException e) {
            System.err.println("Database initialization error: " + e.getMessage());
        }
    }
    
    private static String generateSalt() {
        SecureRandom random = new SecureRandom();
        byte[] salt = new byte[16];
        random.nextBytes(salt);
        return Base64.getEncoder().encodeToString(salt);
    }
    
    private static String hashPassword(String password, String salt) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            md.update(Base64.getDecoder().decode(salt));
            byte[] hashedPassword = md.digest(password.getBytes());
            return Base64.getEncoder().encodeToString(hashedPassword);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Hashing algorithm not found", e);
        }
    }
    
    private static boolean validateName(String name) {
        return name != null && name.trim().length() >= 2 && name.trim().length() <= 100;
    }
    
    private static boolean validateEmail(String email) {
        return email != null && EMAIL_PATTERN.matcher(email).matches();
    }
    
    private static boolean validatePassword(String password) {
        return password != null && password.length() >= 8 && password.length() <= 128;
    }
    
    public static RegistrationResult registerUser(String name, String email, String password) {
        if (!validateName(name)) {
            return new RegistrationResult(false, "Invalid name. Must be 2-100 characters.");
        }
        
        if (!validateEmail(email)) {
            return new RegistrationResult(false, "Invalid email format.");
        }
        
        if (!validatePassword(password)) {
            return new RegistrationResult(false, "Invalid password. Must be 8-128 characters.");
        }
        
        String salt = generateSalt();
        String passwordHash = hashPassword(password, salt);
        
        String sql = "INSERT INTO users (name, email, password_hash, salt) VALUES (?, ?, ?, ?)";
        
        try (Connection conn = DriverManager.getConnection(DB_URL);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            
            pstmt.setString(1, name.trim());
            pstmt.setString(2, email.toLowerCase().trim());
            pstmt.setString(3, passwordHash);
            pstmt.setString(4, salt);
            
            pstmt.executeUpdate();
            return new RegistrationResult(true, "User registered successfully.");
            
        } catch (SQLException e) {
            if (e.getMessage().contains("UNIQUE constraint failed")) {
                return new RegistrationResult(false, "Email already registered.");
            }
            return new RegistrationResult(false, "Registration failed: " + e.getMessage());
        }
    }
    
    public static void main(String[] args) {
        initializeDatabase();
        
        System.out.println("=== User Registration System Test Cases ===\\n");
        
        // Test Case 1: Valid registration
        RegistrationResult result1 = registerUser("John Doe", "john.doe@example.com", "SecurePass123");
        System.out.println("Test 1 - Valid registration: " + result1.message);
        
        // Test Case 2: Duplicate email
        RegistrationResult result2 = registerUser("Jane Doe", "john.doe@example.com", "AnotherPass456");
        System.out.println("Test 2 - Duplicate email: " + result2.message);
        
        // Test Case 3: Invalid email format
        RegistrationResult result3 = registerUser("Bob Smith", "invalid-email", "Password789");
        System.out.println("Test 3 - Invalid email: " + result3.message);
        
        // Test Case 4: Weak password
        RegistrationResult result4 = registerUser("Alice Brown", "alice@example.com", "123");
        System.out.println("Test 4 - Weak password: " + result4.message);
        
        // Test Case 5: Invalid name
        RegistrationResult result5 = registerUser("A", "test@example.com", "ValidPass123");
        System.out.println("Test 5 - Invalid name: " + result5.message);
    }
}
